### PR TITLE
cpm_tpm_rpk

### PR DESCRIPTION
**Pull request by bellenger-l**
cpm_tpm_rpk is a galaxy tool that transforms expression values in CPM, TPM or RPK. It takes two inputs expression matrix and gene length table (for gene length normalization with RPK and TPM).

### CODE REVIEW

File: tools/cpm_tpm_rpk/cpm_tpm_rpk.xml [hunk 1]
```diff
@@ -0,0 +1,100 @@
+<tool id="cpm_tpm_rpk" name="CPM, TPM, RPK" version="0.1.0">
```
Comment (drosofff) [inline tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: name="Generate CPM, TPM and RPK"
<description>from raw count expression values</description>

File: tools/cpm_tpm_rpk/cpm_tpm_rpk.xml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            <param name="gene_file" value="gene_length.tab" ftype="tabular"/>
+            <param name="gene_header" value="FALSE"/>
+            <output name="output" file="tpm.tab" ftype="tabular"/>
+        </test>
+    </tests>
+    <help>
+
+**What it does**
+
+cpm_tpm_rpk takes a table of raw read counts, each column corresponding to a sample count and
```
Comment (drosofff) [inline tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: cpm_tpm_rpk takes a table of raw read counts, first column for gene names and other columns for sample counts,

File: tools/cpm_tpm_rpk/cpm_tpm_rpk.xml [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        </test>
+    </tests>
+    <help>
+
+**What it does**
+
+cpm_tpm_rpk takes a table of raw read counts, each column corresponding to a sample count and 
+the first one is the gene names, and returns a table of transformed expression values either 
+CPM (Count per Million) or TPM (Transcript per million) or RPK (Reads Per Kilobase).
+
```
Comment (drosofff) [inline tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: CPM, TPM and RPK could be better explained

File: tools/cpm_tpm_rpk/cpm_tpm_rpk.xml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+Normalization by gene length (TPM and RPK) require another input file, a two column table of
```
Comment (drosofff) [inline tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: Normalizations by gene lengths (TPM and RPK) require also a two column correspondance table gene_name - gene length

File: tools/cpm_tpm_rpk/cpm_tpm_rpk.xml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+gene length (gene_name - gene length).
+
+Data can also be log-transformed.
```
Comment (drosofff) [inline tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: Not clear

File: tools/cpm_tpm_rpk/cpm_tpm_rpk.xml [hunk 6]
```diff
@@ -87,14 +87,17 @@ Rscript $__tool_directory__/cpm_tpm_rpk.R
 
 **What it does**
 
-cpm_tpm_rpk takes a table of raw read counts, each column corresponding to a sample count and 
-the first one is the gene names, and returns a table of transformed expression values either 
-CPM (Count per Million) or TPM (Transcript per million) or RPK (Reads Per Kilobase).
+cpm_tpm_rpk takes a table of raw read counts, first column is gene name, other corresponding to sample counts and
```
Comment (drosofff) [inline tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: ,first column is gene names and others correspond to sample counts,

File: tools/cpm_tpm_rpk/cpm_tpm_rpk.xml [hunk 7]
```diff
@@ -87,14 +87,17 @@ Rscript $__tool_directory__/cpm_tpm_rpk.R
...
+with RPK (Reads Per Kilobase). The last two manipulations can only apply on full length RNA-sequencing because it
+generates a biais if you normalize gene length with a 3'end sequencing.
 
-Normalization by gene length (TPM and RPK) require another input file, a two column table of
-gene length (gene_name - gene length).
+Normalizations by gene lengths (TPM and RPK) require also a two column correspondance table gene_name - gene length
+where the length must be in nucleotide.
 
-Data can also be log-transformed.
+cpm_tpm_rpk has an option to log-transform results, for example: log2(CPM +1).
```
Comment (drosofff) [inline tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: example ? or log2(TPM+1) is also available ?

Comment (bellenger-l) [reply tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: Yes this is an option either you generate CPM, TPM or RPK or you decide to log transformed them and get log2(CPM+1), log2(TPM+1) or log2(RPK+1)

Comment (bellenger-l) [reply tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: cpm_tpm_rpk permits to log-transform the result of normalization, for example: log2(CPM +1).

Better ?

File: tools/cpm_tpm_rpk/.shed.yml [hunk 8]
```diff
@@ -4,10 +4,12 @@ name: cpm_tpm_rpk
 description: Generate CPM,TPM or RPK from raw counts 
 long_description: |
   Normalizes raw counts expression matrix on different parameters.
-categories:
   - CPM : library-size normalization
   - TPM : gene length and library-size normalization
   - RPK : gene length normalization
+categories:
+  - Sequence Analysis
```
Comment (drosofff) [inline tools/cpm_tpm_rpk/.shed.yml]: Are you sure ?

File: tools/cpm_tpm_rpk/cpm_tpm_rpk.xml [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+			<when value="cpm">
+			</when>
+		</conditional>
+		<param name="log" type="select" label="Data should be log transformed ?">
+			<option value="FALSE" selected="true">No</option>
+			<option value="TRUE">Yes</option>
+		</param>
+	</inputs>
+	<outputs>
+                <data name="output" format="tabular" label="Transformation of raw counts" />
```
Comment (drosofff) [inline tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: In addition to the conditional output, you can add column metadata (that may change in function of the output) using the code (from iuc DESeq2):

https://github.com/galaxyproject/tools-iuc/blob/master/tools/deseq2/deseq2.xml#L164

File: tools/cpm_tpm_rpk/cpm_tpm_rpk.xml [hunk 10]
```diff
@@ -0,0 +1,163 @@
+<tool id="cpm_tpm_rpk" name="Generate CPM, TPM, RPK" version="0.1.0">
+    <description>from raw counts expression values</description>
+    <requirements>
+        <requirement type="package" version="1.6.0=r3.4.2">r-optparse</requirement>
```
Comment (drosofff) [inline tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: R and optparse version are conflicting. use version="`1.3.2=r3.3.2_0`"

File: tools/cpm_tpm_rpk/cpm_tpm_rpk.xml [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        <exit_code range="1:" level="fatal" description="Tool exception" />
+    </stdio>
+    <command detect_errors="exit_code"><![CDATA[ 
+Rscript $__tool_directory__/cpm_tpm_rpk.R 
+-d $input 
+-t $option.type_transfo 
+-s $input_sep 
+-c $input_header 
+#if $option.type_transfo == "tpm":
+-g $option.gene_file
```
Comment (drosofff) [inline tools/cpm_tpm_rpk/cpm_tpm_rpk.xml]: change the name of this option (-- only option ?) to avoid conflicting optparse option -g

**Pull request merged**